staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+	makePod := func(rv uint64) *example.Pod {
+		return &example.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            fmt.Sprintf("pod-%d", rv),
+				Namespace:       "ns",
+				ResourceVersion: fmt.Sprintf("%d", rv),
+				Annotations:     map[string]string{},
+			},
+		}
+	}
+
+	scenarios := []struct {
+		name string
+		run  func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string))
+	}{
+		{
+			name: "allowWatchBookmarks=true, sendInitialEvents=true, RV=unset",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// before starting a new watch set a storage RV to some future value
+				setStorageListMetaRV("102")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{Predicate: func() storage.SelectionPredicate { p := storage.Everything; p.AllowWatchBookmarks = true; return p }(), SendInitialEvents: func() *bool { b := true; return &b }()})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure we only get initial events
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(101)}})
+				verifyNoEvents(t, w)
+				// add a pod that matches the storage's future RV
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				verifyEvents(t, w,
+					[]watch.Event{
+						{Type: watch.Added, Object: makePod(102)},
+						{Type: watch.Bookmark, Object: &example.Pod{
+							ObjectMeta: metav1.ObjectMeta{
+								ResourceVersion: "102",
+								Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+							},
+						}},
+					},
+				)
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "allowWatchBookmarks=true, sendInitialEvents=true, RV=0",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{ResourceVersion: "0", Predicate: func() storage.SelectionPredicate { p := storage.Everything; p.AllowWatchBookmarks = true; return p }(), SendInitialEvents: func() *bool { b := true; return &b }()})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure we only get events up to an RV=102 and the bookmark
+				verifyEvents(t, w,
+					[]watch.Event{
+						{Type: watch.Added, Object: makePod(101)},
+						{Type: watch.Added, Object: makePod(102)},
+						{Type: watch.Bookmark, Object: &example.Pod{
```

Comment:
atm this scenario fails because we don't send a bookmark event immediately, we could change a watcher (processInterval method) or change the scenario.

Reply:
hmm - but verifyEvent is waiting couple seconds so it doesn't have to be immediate.

With realClock(), we should actually send it within a second.
So I don't understand why it is failing.